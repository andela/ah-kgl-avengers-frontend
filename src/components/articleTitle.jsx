import React from 'react';
import PropTypes from 'prop-types';

const ArticleTitle = ({ className, children }) => <h2 className={className}>{children}</h2>;

ArticleTitle.propTypes = {
  className: PropTypes.string.isRequired,
  children: PropTypes.string.isRequired,
};

export default ArticleTitle;
