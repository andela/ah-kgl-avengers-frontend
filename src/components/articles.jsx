import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  TabContent,
  TabPane,
  Row,
  Col,
  Pagination,
  PaginationItem,
  PaginationLink,
} from 'reactstrap';
import { MdDelete, MdEdit } from 'react-icons/md';
import Navbar from './functional/navBar';
import getArticles from '../redux/action-creators/getArticles';
import getDrafts from '../redux/action-creators/getDrafts';
import editArticle from '../redux/action-creators/editArticle';
import DeleteConfirmation from './delete/deleteConfirmation';
import { deleteArticle } from '../redux/action-creators/index';

class Articles extends Component {
  constructor(props) {
    super(props);
    this.toggle = this.toggle.bind(this);

    this.state = {
      activeTab: '1',
      draftClass: false,
      pubClass: true,
      modal: false,
      slug: '',
      status: '',
      currentPage: 0,
      pageSize: 10,
      tab: 1,
    };
  }

  /**
   * check if the user is logged
   */
  componentDidMount() {
    const {
      history,
      loggedIn,
      getArticles: getMyArticles,
      getDrafts: getDraftArticles,
    } = this.props;

    if (typeof loggedIn !== 'undefined' && !loggedIn) return history.push('/');
    getMyArticles();
    return getDraftArticles();
  }

  componentDidUpdate() {
    const { history, loggedIn } = this.props;
    if (typeof loggedIn !== 'undefined' && !loggedIn) return history.push('/');
    return null;
  }

  handleClick = (index) => {
    this.setState({
      currentPage: index,
    });
  };

  toggleDelete = () => {
    const { modal } = this.state;
    this.setState({ modal: !modal });
  };

  onDelete = (slug) => {
    const { deleteArticle: request } = this.props;
    const { status } = this.state;
    request(slug, status);
    this.toggleDelete();
  };

  toggle(tab) {
    const { activeTab, draftClass, pubClass } = this.state;
    if (activeTab !== tab) {
      this.setState({
        activeTab: tab,
        draftClass: !draftClass,
        pubClass: !pubClass,
        currentPage: 0,
      });
    }
  }

  update({ data, status }) {
    const { editArticle: edit, history } = this.props;
    edit(Object.defineProperty(data, 'status', { value: status }));
    history.push('/new-post');
  }

  render() {
    const { articles, drafts } = this.props;

    const {
      activeTab,
      draftClass,
      pubClass,
      currentPage,
      pageSize,
      defaultPages,
      slug,
      modal,
      tab,
    } = this.state;

    const viewArticles = articles.slice(pageSize * currentPage, pageSize * currentPage + pageSize);
    const viewDrafts = drafts.slice(pageSize * currentPage, pageSize * currentPage + pageSize);
    let pages = 1;
    tab === 1 && (pages = Math.ceil(articles.length / pageSize));
    tab === 2 && (pages = Math.ceil(drafts.length / pageSize));

    return (
      <Fragment>
        <div>
          <DeleteConfirmation
            modal={modal}
            toggle={this.toggleDelete}
            deleteArticle={this.onDelete}
            slug={slug}
          />
        </div>
        <Navbar />
        <div className="container articles">
          <h3>My articles</h3>
          <div className="article-menu">
            <button
              type="button"
              onClick={() => {
                this.toggle('1');
                this.setState({ tab: 1 });
              }}
              className={pubClass ? 'bold' : 'normal'}
            >
              Published
            </button>
            <button
              type="button"
              onClick={() => {
                this.toggle('2');
                this.setState({ tab: 2 });
              }}
              className={draftClass ? 'bold' : 'normal'}
            >
              Drafts
            </button>
          </div>
          <div className={articles.length === 0 ? 'no-content' : 'disabled'}>
            <p>No content</p>
          </div>
          <TabContent activeTab={activeTab}>
            {viewArticles
              && viewArticles.map(article => (
                <TabPane tabId="1" key={article.slug}>
                  <Row>
                    <Col className="article col-12">
                      <Link to={`/articles/${article.slug}`}>
                        <h4>{article.title}</h4>
                        <p>{article.description}</p>
                      </Link>
                      <p className="article-details">
                        <span>{article.readTime}</span>
                        <span className="edit">
                          <MdEdit
                            onClick={() => this.update({ data: article, status: 'published' })}
                          />
                        </span>
                        <button
                          type="button"
                          className="delete"
                          onClick={() => {
                            this.setState({ slug: article.slug, status: 'published' });
                            this.toggleDelete();
                          }}
                        >
                          <MdDelete />
                        </button>
                      </p>
                    </Col>
                  </Row>
                  <hr />
                </TabPane>
              ))}
            {viewDrafts
              && viewDrafts.map(draft => (
                <TabPane tabId="2" key={draft.slug}>
                  <Row>
                    <Col className="article col-12">
                      <h4>{draft.title}</h4>
                      <p>{draft.description}</p>
                      <p className="article-details">
                        <span>{draft.readTime}</span>
                        <span className="edit">
                          <MdEdit onClick={() => this.update({ data: draft, status: 'draft' })} />
                        </span>
                        <button
                          type="button"
                          className="delete"
                          onClick={() => {
                            this.setState({ slug: draft.slug, status: 'draft' });
                            this.toggleDelete();
                          }}
                        >
                          <MdDelete />
                        </button>
                      </p>
                    </Col>
                  </Row>
                  <hr />
                </TabPane>
              ))}
            {pages > 1 && (
              <div className="pagination-wrapper">
                <Pagination aria-label="Page navigation example">
                  <PaginationItem disabled={currentPage <= 0}>
                    <PaginationLink
                      onClick={e => this.handleClick(currentPage - 1)}
                      previous
                      href="#"
                    />
                  </PaginationItem>
                  {[...Array(pages || defaultPages)].map((page, i) => (
                    <PaginationItem active={i === currentPage} key={i}>
                      <PaginationLink onClick={() => this.handleClick(i)}>{i + 1}</PaginationLink>
                    </PaginationItem>
                  ))}
                  <PaginationItem disabled={currentPage === pages - 1}>
                    <PaginationLink
                      onClick={() => this.handleClick(currentPage + 1)}
                      next
                      href="#"
                    />
                  </PaginationItem>
                </Pagination>
              </div>
            )}
          </TabContent>
        </div>
      </Fragment>
    );
  }
}

Articles.defaultProps = {
  articles: [],
  drafts: [],
  history: {},
};

Articles.propTypes = {
  getArticles: PropTypes.func.isRequired,
  getDrafts: PropTypes.func.isRequired,
  editArticle: PropTypes.func.isRequired,
  articles: PropTypes.arrayOf(PropTypes.object),
  drafts: PropTypes.arrayOf(PropTypes.object),
  history: PropTypes.objectOf(PropTypes.any),
  deleteArticle: PropTypes.func.isRequired,
  loggedIn: PropTypes.bool.isRequired,
};

const mapStateToProps = ({ article, user: userReducer }) => ({
  articles: article.articles,
  drafts: article.drafts,
  article: article.article,
  loggedIn: userReducer.loggedIn,
});

export default connect(
  mapStateToProps,
  {
    getArticles,
    getDrafts,
    editArticle,
    deleteArticle,
  },
)(Articles);
