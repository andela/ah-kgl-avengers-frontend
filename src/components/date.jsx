import React from 'react';
import PropTypes from 'prop-types';

const ArticleCreatedDate = ({ className, children }) => <h2 className={className}>{children}</h2>;

ArticleCreatedDate.propTypes = {
  className: PropTypes.string.isRequired,
  children: PropTypes.string.isRequired,
};

export default ArticleCreatedDate;
