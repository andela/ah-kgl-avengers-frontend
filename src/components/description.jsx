import React from 'react';
import PropTypes from 'prop-types';

const Description = ({ className, children }) => <p className={className}>{children}</p>;

Description.propTypes = {
  className: PropTypes.string.isRequired,
  children: PropTypes.string.isRequired,
};

export default Description;
