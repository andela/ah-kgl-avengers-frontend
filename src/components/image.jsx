import React from 'react';
import PropTypes from 'prop-types';

const Images = ({ className, alt, image }) => <img className={className} src={image} alt={alt} />;

Images.propTypes = {
  className: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
};

export default Images;
