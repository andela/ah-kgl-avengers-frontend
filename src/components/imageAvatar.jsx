import React from 'react';
import PropTypes from 'prop-types';
import avatar from '../images/avatar.png';

const acronyms = (firstName, lastName) => {
  const names = `${firstName} ${lastName}`;
  if (firstName === '' || lastName === '') {
    return <img className="avatar" src={avatar} alt="avatar" />;
  }
  const acronym = names.match(/\b(\w)/g).join('');
  return <div className="acronym">{acronym}</div>;
};

const ImageAvatar = ({
  image, alt, firstName, lastName, className,
}) => (
  <div className={className}>
    {image ? <img className="avatar" src={image} alt={alt} /> : acronyms(firstName, lastName)}
  </div>
);

ImageAvatar.propTypes = {
  image: PropTypes.string,
  alt: PropTypes.string,
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  className: PropTypes.string,
};

ImageAvatar.defaultProps = {
  image: '',
  alt: '',
  firstName: '',
  lastName: '',
  className: '',
};

export default ImageAvatar;
