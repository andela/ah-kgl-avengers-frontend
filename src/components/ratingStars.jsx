import React from 'react';
import PropTypes from 'prop-types';

const RatingStars = ({ className, children }) => <h2 className={className}>{children}</h2>;

RatingStars.propTypes = {
  className: PropTypes.string,
  children: PropTypes.string,
};

RatingStars.defaultProps = {
  className: '',
  children: null,
};

export default RatingStars;
