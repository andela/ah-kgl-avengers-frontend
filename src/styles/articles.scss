.articles {
  margin-top: 30px;
}

.articles h3{ 
  text-align: center;
}

.articles h4{
  margin-bottom: 20px;
  margin-top: 20px;
  cursor: pointer;
}

.articles hr {
  border-color: #38c4b0;
}

.articles span {
  margin-right: 20px;
  cursor: pointer;
}

.article-details .delete {
  color: #d80909;
  cursor: pointer;
  font-size: 20px;
}
.article-details .edit {
  color: #238100;
  cursor: pointer;
  font-size: 20px;
}

.article {
  cursor: pointer;
}

.article-menu {
  border-bottom: .8px solid #38c4b0;
  padding-bottom: 20px;
  font-size: 20px;
  font-weight: bold;
}

.bold {
  border-bottom: 4px solid #38c4b0;
  padding-bottom: 21px;
}

.normal {
  color: grey;
}

.article-details {
  font-size: 15px;
  color: grey;
}

a {
  text-decoration: none;
  color: black;
}
a:hover {
  text-decoration: none;
  color: black;
}