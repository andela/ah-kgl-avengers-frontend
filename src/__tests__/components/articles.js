import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Articles } from '../../components/articles';

const mockStore = configureMockStore([thunk]);
const store = mockStore({});

const props = {
  getArticles: jest.fn(),
  getDrafts: jest.fn(),
};

describe('Articles', () => {
  const articles = mount(
    <Provider store={store}>
      <Router>
        <Articles {...props} />
      </Router>
    </Provider>,
  );

  it('should render without crashing', () => {
    expect(articles).toMatchSnapshot();
  });

  it('render `my articles` as a title', () => {
    expect(articles.find('h3').text()).toEqual('My articles');
  });
});
