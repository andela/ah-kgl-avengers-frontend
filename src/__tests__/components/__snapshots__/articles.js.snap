// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles should render without crashing 1`] = `ReactWrapper {}`;
